qualtran/bloqs/chemistry/sparse.py

```diff
@@ -0,0 +1,208 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+"""SELECT and PREPARE for the sparse chemistry Hamiltonian in second quantization.
+
+The sparse Hamiltonian simply takes the standard second quantized chemistry
+Hamiltonian and sets to zero any term whose magnitude is smaller than some
+threshold. 
+"""
+
+from functools import cached_property
+from typing import Optional, Set, Tuple, TYPE_CHECKING
+
+import numpy as np
+from attrs import frozen
+from sympy import factorint
+
+from qualtran import Bloq, Register, Signature
+from qualtran.bloqs.basic_gates import TGate
+from qualtran.bloqs.swap_network import CSwapApprox
+
+if TYPE_CHECKING:
+    from qualtran.resource_counting import SympySymbolAllocator
+
+
+@frozen
+class PrepareUniformSuperposition(Bloq):
```

Comment:
What's the naming scheme here? In the THC you have

`UniformSuperpositionTHC` and `PrepareTHC`. The name should imply where in the matrix this bloq resides

Reply:
Should I continue with UniformSuperpositionSparse, etc. They all have potentially slightly different variants of this for each of the algorithms. UniformSuperposition is a subroutine of Prepare typically.

Reply:
I was starting to think that the *Sparse/SF/DF/THC part of the name became somewhat redundant since it's in the sparse module?

Reply:
I'll defer to you for the actual naming scheme; but it should be consistent and explainable.

Relying (solely) on the module to disambiguate may present problems down the line. Elsewhere in qualtran, we use `from module import TheBloq` as a convention because 1) pycharm will add the imports automatically for you 2) I personally find the usages of things more readable if there aren't varying levels of namespacing and abbreviation around them. We could use `sparse.UniformSuperposition` everywhere but might still be easy to mess up. The second reason is that the bloq's name shows up in diagrams. The `short_name` may be shared and ambiguous but it's probably useful for the repr at least to indicate which one is being used.

Reply:
@fdmalone Isn't this exactly same as https://github.com/quantumlib/Qualtran/blob/dc4fecf9d0b761f30d4f896fb49558d8591651d1/qualtran/bloqs/prepare_uniform_superposition.py#L29 ?

Reply:
Oh, here the `d` basis states can by `_any_` `d` basis states and not just the first `d` basis states? The description of the algorithm looks very similar to the what we do for the first `d` basis states, so I would personally want to just generalize the implementation of `PrepareUniformSuperposition` so it takes `d` basis states as input with the default being `[0, d-1]`.

Reply:
It is, but I want to decouple the counts from the implementation as a first pass (see #392). That way I can come back and systematically compare the costs listed in the papers to the implementation. Uniform superposition crops up everywhere in various forms but the costs don't necessarily match those in the literature due to say, inequalities being invertable for free (#390), and so on. Similarly QROAM costs differ from those in the paper (#368), inequalities also differ, rotations cost might differ etc etc. Having a ground "truth" first before implementing a decomposition is easier IMO to track differences.

This way I can come back and substitute out one subbloq at a time with the decomposition and I can more easily track any sources of disagreement. I think this is the best way to go rather than getting bogged down like I did with #350

Reply:
All of that is to say #401 #416 and this will contain redundant code, the main idea is to adopt and correclty partition the costs from openfermion so we can then carefully and systematically make comparisons.

Reply:
Okay, that makes sense. I would suggest we rename it to `UniformSuperpositionSparse` then and open a separate issue (or add a note if there's already an existing issue) documenting that we can probably make incremental changes to `PrepareUniformSuperposition` that can help replace some / all of these new uniform state prep bloqs. 

The incremental change to `PrepareUniformSuperposition`'s implementation can be a good 20% / community contribution task so I think it's worth tracking. 

LGTM for having separate bloqs for now.

Reply:
Cool. Yeah I actually wrote several variations of state preparation when I was getting lost in the weeds in THC (signed state, two registers, and upper triangular (buggy) as cirq circuits) which maybe we could add as a notebook (I think I need your amplitude amplification expertise again).

The lesson here was that it was probably better for me not to do it this way lol. Top down is the way to go, there are too many little details.